src/dead_letters.rs

```diff
@@ -0,0 +1,173 @@
+}
+
+impl DeadLetter {
+    pub fn from_failed_deserialization(bytes: &[u8], err: serde_json::Error) -> Self {
+        let timestamp = Utc::now();
+        Self {
+            base64_bytes: Some(base64::encode(bytes)),
+            json_string: None,
+            error: Some(err.to_string()),
+            timestamp: timestamp.to_rfc3339(),
+            date: timestamp.date().to_string(),
+        }
+    }
+
+    pub fn from_failed_transform(value: &Value, err: TransformError) -> Self {
+        let timestamp = Utc::now();
+        match serde_json::to_string(value) {
+            Ok(s) => Self {
+                base64_bytes: None,
+                json_string: Some(s),
+                error: Some(err.to_string()),
+                timestamp: timestamp.to_rfc3339(),
+                date: timestamp.date().to_string(),
+            },
+            _ => unreachable!(),
+        }
+    }
+
+    pub fn from_failed_parquet_row(value: &Value, err: ParquetError) -> Self {
+        let timestamp = Utc::now();
+        match serde_json::to_string(value) {
+            Ok(s) => Self {
+                base64_bytes: None,
+                json_string: Some(s),
+                error: Some(err.to_string()),
+                timestamp: timestamp.to_rfc3339(),
+                date: timestamp.date().to_string(),
+            },
+            _ => unreachable!(),
+        }
+    }
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum DeadLetterQueueError {
+    #[error("JSON serialization failed: {source}")]
+    SerdeJson {
+        #[from]
+        source: serde_json::Error,
+    },
+
+    #[error("Delta write failed: {source}")]
+    DeltaWriter {
+        #[from]
+        source: DeltaWriterError,
+    },
+}
+
+pub struct DeadLetterQueueOptions {
+    /// Table URI of the delta table to write dead letters to. Implies usage of the DeltaSinkDeadLetterQueue.
+    pub delta_table_uri: Option<String>,
+}
+
+#[async_trait]
+pub trait DeadLetterQueue: Send + Sync {
+    async fn write_dead_letter(
+        &mut self,
+        dead_letter: DeadLetter,
+    ) -> Result<(), DeadLetterQueueError> {
+        self.write_dead_letters(vec![dead_letter]).await
+    }
+
+    async fn write_dead_letters(
+        &mut self,
+        dead_letters: Vec<DeadLetter>,
+    ) -> Result<(), DeadLetterQueueError>;
+}
+
+pub async fn dlq_from_opts(
+    options: DeadLetterQueueOptions,
+) -> Result<Box<dyn DeadLetterQueue>, DeadLetterQueueError> {
+    if let Some(table_uri) = options.delta_table_uri {
+        Ok(Box::new(
+            DeltaSinkDeadLetterQueue::for_table_uri(table_uri.as_str()).await?,
+        ))
+    } else {
+        Ok(Box::new(NoopDeadLetterQueue {}))
+    }
+}
+
+pub struct NoopDeadLetterQueue {}
+
+#[async_trait]
+impl DeadLetterQueue for NoopDeadLetterQueue {
+    async fn write_dead_letters(
+        &mut self,
+        _dead_letters: Vec<DeadLetter>,
+    ) -> Result<(), DeadLetterQueueError> {
+        // noop
+        Ok(())
+    }
+}
```

Comment:
For what purpose does this Noop thing exist?

Reply:
added some rust docs in a recent commit to clarify, but noop is used as the default for folks who dont want to setup extra infra for capturing dead letters. i didnt want to make the logging one default because i dont want to create unexpected pii exposure in logs. having the noop impl there simplifies the caller context a bit so we dont have to unpack an option.

Reply:
+1 on Noop, additional infra / resources should be a choice

Reply:
But that's also interesting, it's a clear example of using polymorphism as a tool. The question for rust experts (cc @houqp) is this a preferred way. Or just simply put it into option and add `if` block on use case

Reply:
fwiw and YAGNI aside - I think other DeadLetterQueue implementations will be highly desirable (i.e. Kafka topic - which we almost implemented) which is the primary reason I chose to stick this behind a trait.